README.md
[![awscdk-jsii-template](https://img.shields.io/badge/built%20with-awscdk--jsii--template-blue)](https://github.com/pahud/awscdk-jsii-template)
[![NPM version](https://badge.fury.io/js/cdk-fargate-fastautoscaler.svg)](https://badge.fury.io/js/cdk-fargate-fastautoscaler)
[![PyPI version](https://badge.fury.io/py/cdk-fargate-fastautoscaler.svg)](https://badge.fury.io/py/cdk-fargate-fastautoscaler)
![Build](https://github.com/aws-samples/aws-fargate-fast-autoscaler/workflows/Build/badge.svg)

## aws-fargate-fast-autoscaler

**AWS Fargate Fast Autosaler** - A Serverless Implementation that Triggers your AWS Fargate autoscaling in seconds with `cdk-fargate-fastautoscaler`.

## cdk-fargate-fastautoscaler

`cdk-fargate-fastautoscaler` is a [aws/jsii](https://github.com/aws/jsii) construct library for AWS CDK.

By building your AWS CDK stacks with `cdk-fargate-fastautoscaler`, you can create your customized Fargate workload with the fast autoscaling capabilities.

![](images/fargate-fast-autoscaler.png)


# How it works

Behind the scene, our workload in PHP, NodeJS, Java or Python is running with a nginx reverse proxy within a single AWS Fargate Task exposing a `/nginx_status` endpoint for realtime connections info generation. All traffic coming through ALB to Fargate tasks will establish active connecitons with the nginx reverse proxy before it can hit our backend server. 

We are running an AWS Step Function state machine to periodically invoke the AWS Lambda function and collect active connection numbers from each Fargate Task **every 3 seconds** and determine our scaling policy in the state machine followed by immediate `ecs service update` to increase the desired number of Fargate tasks.

# AWS CDK Sample

The following CDK sample creates a PHP service in AWS Fargate with the nginx as the reverse proxy.

```ts
import * as cdk from '@aws-cdk/core'
import * as ec2 from '@aws-cdk/aws-ec2';
import { AwsLogDriver, ContainerImage } from '@aws-cdk/aws-ecs';
import { FargateFastAutoscaler } from 'cdk-fargate-fastautoscaler';
import * as path from 'path';

const app = new cdk.App()

const env = {
  region: process.env.CDK_DEFAULT_REGION,
  account: process.env.CDK_DEFAULT_ACCOUNT,
};

const stack = new cdk.Stack(app, 'FargateFastAutoscalerDemo', { env })

const vpc = ec2.Vpc.fromLookup(stack, 'Vpc', { isDefault: true })

new FargateFastAutoscaler(stack, 'FargateFastAutoscaler', {
  vpc,
  // create the backend PHP service
  backendContainer: {
    image: ContainerImage.fromAsset(path.join(__dirname, '../../sample/backend/php')),
  },
  // PHP service running on container port 2015
  backendContainerPortMapping: [ { containerPort: 2015 } ],
})
```

On deployment complete, you'll see the URL in the Outputs:

**fargate-fast-autoscaling.URL** = http://farga-exter-1GW64WGQYNE4O-1567742142.us-west-2.elb.amazonaws.com

Open this URL and you will see the Caddy web server welcome page with phpinfo.

![](images/php-welcome.png)


And if you append `/nginx_status` in the URL and reload the page, you'll see this page:

![](images/nginx-status.png)

### Start your state machine

Go to Step Function console and click **start execution** on the state machines. Leave the execution name and input column as is and click **start execution** again. Your state machine will be running. Behind the scene the step function will invoke a Lambda function to collect **Active Connections** number from nginx reverse proxy on each fargate task and determine a new desired number of fargate tasks to scale. Typically it would just take **less than 10 seconds** before it starts to scale.

![](images/stepfunc.png)

# SNS Service Integration

The **SNSScaleOut** task in the state machine leverages direct Amazon SNS service integration to publish a notification to your SNS topic. You will receive SNS notification when it starts **ServiceScaleOut** task.

Specify the `snsTopic` property to define your custom SNS topic. If not defined, the construct will create a default SNS topic.


# Disable Scale In

By default, `disableScaleIn` is set to true to prevent your workload from scale-in. If you prefer to enable scale in, set `disableScaleIn` to `false`.


## License Summary

This sample code is made available under the MIT-0 license. See the LICENSE file.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.2.48](https://github.com/aws-samples/aws-fargate-fast-autoscaler/compare/v0.2.47...v0.2.48) (2021-04-01)


### Bug Fixes

* bump xmldom ([#246](https://github.com/aws-samples/aws-fargate-fast-autoscaler/issues/246)) ([b60f13e](https://github.com/aws-samples/aws-fargate-fast-autoscaler/commit/b60f13e09a98c0d3cb00bf8786210d4de2b20a25))
* bump xmldom ([#248](https://github.com/aws-samples/aws-fargate-fast-autoscaler/issues/248)) ([cb5386d](https://github.com/aws-samples/aws-fargate-fast-autoscaler/commit/cb5386d9709e64b6bb46d6ae3f9d2031baf9b4c2))

### [0.2.47](https://github.com/aws-samples/aws-fargate-fast-autoscaler/compare/v0.2.46...v0.2.47) (2021-02-23)

### [0.2.46](https://github.com/aws-samples/aws-fargate-fast-autoscaler/compare/v0.2.42...v0.2.46) (2021-02-17)


### Bug Fixes

* add defaultReleaseBranch ([370a79c](https://github.com/aws-samples/aws-fargate-fast-autoscaler/commit/370a79c4441263671db7a258ea213ba25f9033e0))

### [0.2.45](https://github.com/aws-samples/aws-fargate-fast-autoscaler/compare/v0.2.42...v0.2.45) (2021-02-17)


### Bug Fixes

* add defaultReleaseBranch ([370a79c](https://github.com/aws-samples/aws-fargate-fast-autoscaler/commit/370a79c4441263671db7a258ea213ba25f9033e0))

### 0.2.44 (2020-12-19)

### 0.2.43 (2020-12-19)

### 0.2.42 (2020-10-17)

### 0.2.41 (2020-10-16)

### 0.2.40 (2020-10-05)

### 0.2.39 (2020-10-05)

### 0.2.38 (2020-10-03)

### 0.2.37 (2020-09-30)

### 0.2.36 (2020-09-30)

### 0.2.35 (2020-09-30)

### 0.2.34 (2020-09-30)

### 0.2.33 (2020-09-30)

### 0.2.32 (2020-09-30)

### 0.2.31 (2020-09-30)

### 0.2.30 (2020-09-30)

### 0.2.29 (2020-09-30)

### 0.2.28 (2020-09-28)

### 0.2.27 (2020-09-23)

### 0.2.26 (2020-09-23)

### 0.2.25 (2020-09-22)

### [0.2.24](https://github.com/aws-samples/aws-fargate-fast-autoscaler/compare/v0.2.23...v0.2.24) (2020-09-09)

### [0.2.23](https://github.com/aws-samples/aws-fargate-fast-autoscaler/compare/v0.2.22...v0.2.23) (2020-09-08)

### [0.2.22](https://github.com/aws-samples/aws-fargate-fast-autoscaler/compare/v0.2.20...v0.2.22) (2020-09-06)

### 0.2.21 (2020-09-01)

### [0.2.20](https://github.com/pahud/cdk-spot-one/compare/v0.2.19...v0.2.20) (2020-09-01)

### 0.2.19 (2020-08-29)

### 0.2.18 (2020-08-28)

### 0.2.17 (2020-08-28)

### 0.2.16 (2020-08-28)

### [0.2.15](https://github.com/pahud/cdk-spot-one/compare/v0.2.14...v0.2.15) (2020-08-28)

### [0.2.14](https://github.com/pahud/cdk-spot-one/compare/v0.2.13...v0.2.14) (2020-08-28)

### 0.2.13 (2020-08-28)

### [0.2.12](https://github.com/pahud/cdk-spot-one/compare/v0.2.11...v0.2.12) (2020-08-27)

### [0.2.11](https://github.com/pahud/cdk-spot-one/compare/v0.2.8...v0.2.11) (2020-08-27)

### [0.2.9](https://github.com/pahud/cdk-spot-one/compare/v0.2.7...v0.2.9) (2020-08-24)

### 0.2.10 (2020-08-27)

### 0.2.9 (2020-08-24)

### 0.2.8 (2020-08-24)

### [0.2.7](https://github.com/aws-samples/aws-fargate-fast-autoscaler/compare/v0.2.6...v0.2.7) (2020-08-24)

### 0.2.6 (2020-08-19)

### 0.2.5 (2020-08-19)

### [0.2.4](https://github.com/aws-samples/aws-fargate-fast-autoscaler/compare/v0.2.3...v0.2.4) (2020-08-15)

### [0.2.3](https://github.com/aws-samples/aws-fargate-fast-autoscaler/compare/v0.2.2...v0.2.3) (2020-08-13)

### [0.2.2](https://github.com/aws-samples/aws-fargate-fast-autoscaler/compare/v0.2.1...v0.2.2) (2020-08-13)

### 0.2.1 (2020-08-08)

## [0.2.0](https://github.com/aws-samples/aws-fargate-fast-autoscaler/compare/v0.1.2...v0.2.0) (2020-08-08)

### [0.1.2](https://github.com/aws-samples/aws-fargate-fast-autoscaler/compare/v0.1.1...v0.1.2) (2020-08-08)

### [0.1.1](https://github.com/aws-samples/aws-fargate-fast-autoscaler/compare/v0.1.0...v0.1.1) (2020-08-08)

### [0.1.1](https://github.com/aws-samples/aws-fargate-fast-autoscaler/compare/v0.1.0...v0.1.1) (2020-08-08)

## 0.1.0 (2020-08-08)

## 0.1.0 (2020-08-08)


cdk/lib/cdk-stack.ts
import cdk = require("@aws-cdk/core");
import ec2 = require("@aws-cdk/aws-ec2");
import ecs = require("@aws-cdk/aws-ecs");
import elbv2 = require("@aws-cdk/aws-elasticloadbalancingv2");
import iam = require("@aws-cdk/aws-iam");
import sfn = require('@aws-cdk/aws-stepfunctions');
import sfn_tasks = require('@aws-cdk/aws-stepfunctions-tasks');
import lambda = require('@aws-cdk/aws-lambda');
import sns = require('@aws-cdk/aws-sns');

import { CfnResource } from "@aws-cdk/core";


const DEFAULT_SNS_TOPIC_ARN = 'arn:aws:sns:ap-northeast-1:903779448426:SNS2IM'
const AWSCLI_LAYER_ARN = 'arn:aws:serverlessrepo:us-east-1:903779448426:applications/lambda-layer-awscli'
const AWSCLI_LAYER_VERSION = '1.16.232'

export class AwsFargateFastAutoscalerStack extends cdk.Stack {
  public readonly fargateWatcherFuncArn: string;
  public readonly layerVersionArn: string;
  public readonly fargateService: ecs.FargateService;
  public readonly fargateTaskDef: ecs.FargateTaskDefinition

  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // import default VPC
    const vpc = ec2.Vpc.fromLookup(this, 'VPC', {
      isDefault: true
    });

    // create a security group that allows all traffic from the same sg
    const sg = new ec2.SecurityGroup(this, 'SharedSecurityGroup', {
      vpc: vpc,
    })
    sg.connections.allowFrom(sg, ec2.Port.allTraffic())


    //sg for HTTP public access
    const httpPublicSecurityGroup = new ec2.SecurityGroup(this, "HttpPublicSecurityGroup", {
      allowAllOutbound: true,
      securityGroupName: 'HttpPublicSecurityGroup',
      vpc: vpc
    });

    httpPublicSecurityGroup.connections.allowFromAnyIpv4(ec2.Port.tcp(80));


    // // Fargate Cluster
    const fgCluster = new ecs.Cluster(this, 'fgCluster', {
      vpc: vpc
    });


    // // task iam role
    const taskIAMRole = new iam.Role(this, 'fgDemoTaskExecutionRole', {
      assumedBy: new iam.ServicePrincipal('ecs-tasks.amazonaws.com')
    });

    taskIAMRole.addToPolicy(new iam.PolicyStatement({
      resources: ['*'],
      actions: ['xray:PutTraceSegments']
    }));

    // ECS task definition
    const demoTaskDef = new ecs.FargateTaskDefinition(this, 'cdk-fargate-demo-taskdef', {
      cpu: 256,
      memoryLimitMiB: 512,
      taskRole: taskIAMRole,
    })

    this.fargateTaskDef = demoTaskDef

    const mainContainer = demoTaskDef.addContainer('main', {
      image: ecs.ContainerImage.fromAsset('./nginx', {}),
      cpu: 0,
      logging: new ecs.AwsLogDriver({
        streamPrefix: 'echo-http-req'
      }),
    })

    const phpContainer = demoTaskDef.addContainer('backend', {
      image: ecs.ContainerImage.fromAsset('./php', {}),
      cpu: 0,
      logging: new ecs.AwsLogDriver({
        streamPrefix: 'echo-http-req'
      })
    })

    //   mainContainer.addLink(phpContainer, 'app')
    mainContainer.addPortMappings({
      containerPort: 80
    })

    phpContainer.addPortMappings({
      containerPort: 2015
    })

    const demoService = new ecs.FargateService(this, 'demo-service', {
      cluster: fgCluster,
      desiredCount: 2,
      taskDefinition: demoTaskDef,
      securityGroup: sg,
    });

    this.fargateService = demoService


    const externalLB = new elbv2.ApplicationLoadBalancer(this, 'external', {
      vpc: vpc,
      internetFacing: true,
      securityGroup: httpPublicSecurityGroup,
    });

    const externalListener = externalLB.addListener('PublicListener', {
      port: 80
    });


    const healthCheckDefault = {
      "port": 'traffic-port',
      "path": '/',
      "intervalSecs": 30,
      "timeoutSeconds": 5,
      "healthyThresholdCount": 5,
      "unhealthyThresholdCount": 2,
      "healthyHttpCodes": "200,301,302"
    };

    externalListener.addTargets('fg-echo-req', {
      port: 80,
      protocol: elbv2.ApplicationProtocol.HTTP,
      healthCheck: healthCheckDefault,
      targets: [demoService],
      deregistrationDelay: cdk.Duration.seconds(3)
    });


    // SAM
    // IAM role for lambda
    const lambdaRole = new iam.Role(this, 'lambdaRole', {
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com')
    });
    lambdaRole.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonECS_FullAccess'))
    lambdaRole.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonEC2ReadOnlyAccess'))

    lambdaRole.addToPolicy(new iam.PolicyStatement({
      resources: ['*'],
      actions: [
        'ec2:CreateNetworkInterface',
        'ec2:DescribeNetworkInterfaces',
        'ec2:DeleteNetworkInterface',
      ]
    }));


    lambdaRole.addToPolicy(new iam.PolicyStatement({
      resources: ['*'],
      actions: [
        'logs:CreateLogGroup',
        'logs:CreateLogStream',
        'logs:PutLogEvents',
      ]
    }));

    // const uniqueId = crypto.createHash('md5').update(this.node.path).digest("hex");
    this.templateOptions.transforms = ['AWS::Serverless-2016-10-31']; // required for AWS::Serverless
    const resource = new CfnResource(this, 'Resource', {
      type: 'AWS::Serverless::Application',
      properties: {
        Location: {
          ApplicationId: AWSCLI_LAYER_ARN,
          SemanticVersion: AWSCLI_LAYER_VERSION
        },
        Parameters: {}
      }
    })
    this.layerVersionArn = cdk.Token.asString(resource.getAtt('Outputs.LayerVersionArn'));

    const fargateWatcherFunc = new lambda.Function(this, 'fargateWatcherFunc', {
      runtime: lambda.Runtime.PROVIDED,
      handler: 'main',
      code: lambda.Code.fromAsset('./sam/fargateWatcherFunc/func.d'),
      layers: [lambda.LayerVersion.fromLayerVersionArn(this, 'AwsCliLayer', this.layerVersionArn)],
      memorySize: 1024,
      timeout: cdk.Duration.minutes(1),
      role: lambdaRole,
      vpc: vpc,
      vpcSubnets: { subnetType: ec2.SubnetType.PRIVATE },
      securityGroup: sg,
      environment: {
        cluster: fgCluster.clusterName,
        service: demoService.serviceName,
        disable_scalein: 'yes',
        region: `${this.region}`
      },
    });

    this.fargateWatcherFuncArn = fargateWatcherFunc.functionArn

    // step function
    const wait3 = new sfn.Wait(this, 'Wait 3 Seconds', {
      // time: sfn.WaitTime.secondsPath('$.wait_time') 
      time: sfn.WaitTime.duration(cdk.Duration.seconds(3))
    });
    const wait60 = new sfn.Wait(this, 'Wait 60 Seconds', {
      time: sfn.WaitTime.duration(cdk.Duration.seconds(60))
    });

    const getEcsTasks = new sfn.Task(this, 'GetECSTasks', {
      task: new sfn_tasks.InvokeFunction(lambda.Function.fromFunctionArn(this, 'getEcsTasks', fargateWatcherFunc.functionArn)),
      resultPath: '$.status',
    });

    const snsScaleOut = new sfn.Task(this, 'SNSScaleOut', {
      task: new sfn_tasks.PublishToTopic(sns.Topic.fromTopicArn(this, 'snsTopic', DEFAULT_SNS_TOPIC_ARN), {
        // message: sfn.TaskInput.fromDataAt('$'),
        message: sfn.TaskInput.fromObject({
          'Input.$': '$'
        }),
        subject: "Fargate Start Scaling Out",
      }),
      resultPath: "$.taskresult",

    });
    const svcScaleOut = new sfn.Task(this, 'ServiceScaleOut', {
      task: new sfn_tasks.InvokeFunction(lambda.Function.fromFunctionArn(this, 'svcScaleOut', fargateWatcherFunc.functionArn)),

    });

    const isServiceOverloaded = new sfn.Choice(this, 'IsServiceOverloaded', {
      inputPath: '$.status'
    });
    const isDone = new sfn.Pass(this, 'Done')

    const desire2 = new sfn.Pass(this, 'Desire2', {
      outputPath: '$',
      result: sfn.Result.fromObject({ Desired: 2 })
    })
    // const desire5 = new sfn.Pass(this, 'Desire5', {
    //     outputPath: DISCARD,
    //     result: sfn.Result.fromObject({Desired: 5})
    // })
    const desire10 = new sfn.Pass(this, 'Desire10', {
      outputPath: '$',
      result: sfn.Result.fromObject({ Desired: 10 })
    })
    const desire15 = new sfn.Pass(this, 'Desire15', {
      outputPath: '$',
      result: sfn.Result.fromObject({ Desired: 15 })
    })
    const desire20 = new sfn.Pass(this, 'Desire20', {
      outputPath: '$',
      result: sfn.Result.fromObject({ Desired: 20 }),
    })

    const chain = sfn.Chain
      .start(getEcsTasks)
      .next(isServiceOverloaded
        .when(sfn.Condition.numberGreaterThanEquals('$.avg', 500), desire20
          .next(snsScaleOut
            .next(svcScaleOut
              .next(wait60
                .next(getEcsTasks
                )))))
        .when(sfn.Condition.numberGreaterThanEquals('$.avg', 300), desire15
          .next(snsScaleOut
          ))
        .when(sfn.Condition.numberGreaterThanEquals('$.avg', 100), desire10
          .next(snsScaleOut
          ))
        .when(sfn.Condition.numberGreaterThanEquals('$.avg', 50), desire2
          .next(snsScaleOut
          ))
        // .when(sfn.Condition.numberLessThanEquals('$.avg', 10), desire2
        //     .next(snsScaleOut
        // ))
        .when(sfn.Condition.numberLessThan('$.avg', 0), isDone)
        .otherwise(wait3
          .next(getEcsTasks)
        ));

    new sfn.StateMachine(this, 'FargateFastAutoscaler', {
      definition: chain,
      timeout: cdk.Duration.hours(24)
    });

    new cdk.CfnOutput(this, 'ClusterARN: ', { value: fgCluster.clusterArn });
    new cdk.CfnOutput(this, 'URL: ', { value: 'http://' + externalListener.loadBalancer.loadBalancerDnsName });
    new cdk.CfnOutput(this, 'FargateWatcherLambdaArn: ', { value: fargateWatcherFunc.functionArn });

  }

}
